Add data loader to database

pasan04 opened the pull request:
Here, the data loading to the database which reads the parquet files is added.

Commit:
add data loading to database

mr-devs commented on data-loader/server.py:
```diff
@@ -0,0 +1,56 @@
+                read_dir = twitter_data_root
+            controller.read_files(read_dir, platform, selected_month)
+
+def get_available_year_month(root_dir):
```
It doesn't look like this function is used in this script. Can we remove it?

pasan04 replied:
I removed the get_avaialble_year_month function. As there is no need to search for available months in the directory, I can call the current_year_month directory without searching for the entire directory

mr-devs commented on data-loader/server.py:
```diff
@@ -0,0 +1,56 @@
+PLATFORMS = ["Facebook", "Twitter"]
+LOAD_PAST_MONTH_DATA = ['2023_01']
+
+def load_current_data():
```
It doesn't look like this function is used in this script. Can we remove it? If you need it for something else, I would either put it in that script or create a separate .py file to hold these functions.

pasan04 replied:
This function is used to get the current year month along with the directory. so in here, the it will pass the platform, current month directory to controller. The load_past_months_data function is used to add the past months data (ex- 2022_01, 2022_02, etc ) to the database.

mr-devs commented on data-loader/database_functions/fib_indices.py:
```diff
@@ -0,0 +1,44 @@
+import logging
+from flask import Flask
+from library import backend_util
+logger = logging.getLogger(__name__)
```
For all of the scripts, can you please utilize the `get_logger` function that I have built into the package? See [here](https://github.com/mr-devs/top-FIBers/blob/e0e60f99f628b04b66910cb158ec550d448dd571/package/top_fibers_pkg/utils.py#L249).

This will ensure that we are logging everything in the same way. Also, as you're using the logger right now, (1) I don't believe that the logging messages are being recorded in a file that we'll be able to reference later. We will need this for when everything is automated by a cronjob. 

And (2) I don't think that the `logger.info` messages are being delivered (as the default logging level is WARNING [aka 30]). You can check this by sticking `print(logger.getEffectiveLevel())` right after this line. either way I think we should be consistent across the project.

pasan04 replied:
I added the logger to read the files and the output file will be in /home/data/apps/topfibers/repo/logs

mr-devs commented on data-loader/server.py:
```diff
@@ -0,0 +1,56 @@
+    """
+    Load each platform and read the files.
+    """
+    for selected_month in LOAD_PAST_MONTH_DATA:
```
It looks like `LOAD_PAST_MONTH_DATA` is a list of one string. Is this what we want? it seems like we would need to update this manually each month. This is not what we want as it seems like it will defeat the purpose of the cronjob. Shouldn't we instead want to just automatically create the `YYYY_mm` string for the past month with something like `datetime`?

pasan04 replied:
This LOAD_PAST_MONTH data is used to upload the previous months data as an array. (2022_01,2022_02 ... to 2023_01), So I kept this, because there is another month remained to upload for the 2022_02 data.

mr-devs commented on data-loader/run_data_loader.sh:
```diff
@@ -0,0 +1,18 @@
+ssh -i /u/truthy/.ssh/id_rsa -4 -N -L 5580:localhost:5432 truthy@lisa.luddy.indiana.edu &
+
+# Store the PID of the SSH tunnel process
+tunnel_pid=$!
```
ooh what does `$!` do?

I can't find anything about this?

What I know if is:

| Variable | Use |
| -- | -- |
| $# | Stores the number of command-line arguments that were passed to the shell program. |
| $? | Stores the exit value of the last command that was executed. |
| $0 | Stores the first word of the entered command (the name of the shell program). |
| $* | Stores all the arguments that were entered on the command line ($1 $2 ...). |
| "$@" | Stores all the arguments that were entered on the command line,    individually quoted ("$1" "$2" ...). |

from http://linuxsig.org/files/bash_scripting.html

pasan04 replied:
I am not sure this is the best method, but this works as $! will keep the background process id. Here some links, that it has, https://www.cyberciti.biz/faq/how-to-return-pid-of-a-last-command-in-linux-unix/ ,  https://stackoverflow.com/questions/29340087/bash-store-command-pid-in-variable-and-kill-process
Any suggestions to make this..

mr-devs reviewed (changes requested):
This looks great. Thank you for all of this work! I definitely did not realize how much went into setting up the database.

I left a few comments in-line but most of my feedback has to do with documentation and organization. Here is what I would suggest:
1. Add README.md files for all directories with a high-level description of each file in the directory
2. Add docstrings for all of the scripts 
    - See [iffy_update.py](https://github.com/mr-devs/top-FIBers/blob/e0e60f99f628b04b66910cb158ec550d448dd571/scripts/data_collection/iffy_update.py) for a python example and [iffy_get_data.sh](https://github.com/mr-devs/top-FIBers/blob/e0e60f99f628b04b66910cb158ec550d448dd571/scripts/data_collection/iffy_get_data.sh) for a bash example. The general idea here is to capture the purpose of the script (w. overview of how it works), inputs, outputs, and author (so people know who to reach out to for questions).
4. Add docstrings for all functions
    - You can check any of the scripts that I wrote for an example, [here](https://github.com/mr-devs/top-FIBers/blob/e0e60f99f628b04b66910cb158ec550d448dd571/scripts/data_prep/create_data_file_symlinks.py) is the first one that I came across. General idea is to capture inputs and outputs (as well as their types) as well as exceptions
5. I would consider adding some more high-level comments to the code for clarity around what certain blocks are accomplishing. I know that most programmers really hate commenting their code but it can be really helpful for the future. E.g. `controller.py` has a lot going on and zero comments. At the end of the day, it is fairly simple but digesting all of this code at once is really overwhelming and it would've been nice to have a bit of guidance in plain english.
6. In many of the scripts you have some constants for RELATIVE paths. I am sure that you had to update all of these once I asked you to move the directory! Instead, just make them absolute paths, that way it does not matter if we decide to save this code somewhere else. Actually, this seems fairly likely at this stage, IMO. 
7. It looks like there is still a `requirements.txt` file. From what you told me, all of this code is run with the existing conda environment so I think we can remove this?
9. It looks like there is a `.gitignore` file. We also will not need this anymore. If anything is not included in the larger repo's `.gitignore` please add it there.
10. `test_posts.py` is an empty file. Can we delete it?

At the end of the day, all of the above is just a suggestion as this code will not be managed by me. However, I know we have been trying to future-proof things a lot lately so I would strongly recommend it.

Commit:
add readme files

Commit:
added directory README files

Commit:
add changes to readme files

Commit:
add documentation to each file

Commit:
add code review changes

Commit:
add code review changes

Commit:
Add the log files and remove unwanted logs in the log files

pasan04 commented:
I have corrected the issues mentioned. can you recheck this.

Commit:
Clean up README files

Commit:
Update script documentation, formatting (black), etc.

mr-devs commented:
@pasan04 Thanks so much for making the latest commits about the documentation. I went through things a bit myself and committed some changes. 

Before going into those changes, a couple of things:
1. No matter what, I think that we can keep the changes I made in the commit whose message is ["Clean up README files"](32aa63ac24a87343d495711989a3584583672e14).
2. However, **please check** the changes I've made (summarized below) in the commit whose message is ["Update script documentation, formatting (black), etc."](890796000c3d516a2437ce7bb1574a2a153ab52f). **If for some reason I've messed things up so bad that you want to just throw away the changes I've made, you can pull down the branch that I've made and `reset --hard` the log so that only the readme changes have been made. To do this, you would run**:
```
git reset --hard 32aa63ac24a87343d495711989a3584583672e14
```

> `32aa63ac24a87343d495711989a3584583672e14` can be found by running `git log` and looking at the commit code above the "Clean up README files" message.

After that, you just would push the changes to the branch.

---

Apologies for the length. If anything isn't clear, please let me know if you'd like to go through it together.

### My changes
- I updated the `README.md` files so that they were a bit more uniform in their structure across the different directories. I believe now they all have a `Contents` section and then basically list the files and what they do. Other necessary information is included where necessary. You'll also see that I added a couple of new README files that were missing.
- I cleaned up the docstrings for the scripts and functions where necessary. Previously you had the inputs for _functions_ listed in the inputs section for the _script_, and did not have docstrings for most of the functions. I have added them with the standard triple quotes format (`"""docstring example"""`) under where the function is defined. Aside from the clarity that they add to the code, other systems are built expecting docstrings to be there and they know how to work with that format. For example, in VS code, if I am in script `example.py` and I hover my mouse over a function in that script which is saved in another file called `functions.py`, it will show me tooltip of that functions docstring — so you have less flipping back and forth between scripts to understand what everything does. Also, if start a Python session in your terminal you can run `help(some_function)` and it will print the docstring for `some_function`, again, saving you the hassle of having to look through code.
    - **Please make sure to go through these docstrings and make sure that I did not mess anything up**
- Generally, all files that I edited were updated to use the [`black`](https://dev.to/adamlombard/how-to-use-the-black-python-code-formatter-in-vscode-3lo0) formatting style because my VS code automatically formats python scripts this way. Most of these things update very long lines to use multiple lines instead.
- I made some changes to code that I felt made things a bit cleaner:
    - `server.py`: I removed the two functions you had `load_current_month_data()` and `load_past_months_data()`. The code only ran the `load_past_months_data()` function which was actually set up to iterate through `LOAD_PAST_MONTH_DATA = ['2022_01']` which is confusing because you can pass anything to that list (not just the past month). Now there is only one function called `update_database()` (I think this name is more specific) which iterates through a list called `MONTHS`. You'll see that by default it runs the current month but you can comment in other options for specific months or all months.
        - Instead of using the existing `get_logger` function in `top_fibers_pkg.utils`, you had copied the function over into a new `utils.py` script. There is no need for this so I loaded the function from `top_fibers_pkg.utils`. However, I just realized that I did not delete your duplicate `utils.py` script. **Can you please delete it?**
    - `fib_indices.py`: there is a function in this script (`get_all_posts()`) that looks like it should be deleted. Can you confirm and then please delete it? Given the other `database_functions` code, it also seems like perhaps you would like to add a function that retrieves all of the data from the FIB indices table? If not, please ignore.
    - `controller.py`: I cleaned up the comments and some of the logging messages in this one. For example, the first line in the `read_files()` function (now called `add_data()` as this seems more specific) logged a message that said `"Found the data directory on that day!"` before that had actually happened in the code. Another piece of code that I think can be improved has to do with the `reports.check_report_already_added()` function (however, there is a chance I am misunderstanding it, which is why I did not change it directly). I am wondering if you can flip the boolean value that is returned. Right now, it seems like it returns `False` if a report exists and `True` if it does not. So when you actually use the function like `reports.check_report_already_added(file_date, selected_month, platform)` it seems like it is asking the question "has a report been added for `selected_month`?" and returning `False` when it exists is sort of backward. If my assumption about the output is correct — i.e., it outputs `False` when a report exists — I would suggest that you change the function name to `report_already_added()` and have it return `True` if the report exists and `False` if not. Then the if block for the `controller.py` script will be changed to...
        ```py
        if not reports.report_already_added(file_date, selected_month, platform):
            # add the new data
        ```
    - ... which then reads like "if this report does not exist, we then add the data" which is much more natural.
- There were some packages there were imported here and there which were not used, these imports were removed.

Commit:
Add code review changes

Commit:
removed unused root_dir

pasan04 commented:
Thanks Matt for the code review. You have done an excellent job to make the code more readable and also removed the unwanted code snippets and organize the code well and also I made the corrections in the code as you suggested.  I removed the `root_dir=..` from the controller.py as it is not used.

Commit:
change the log file path and bug fixed in database configuration path

Commit:
fixed the database config issue

Commit:
change a logger text

Commit:
add the logger  to controller and utils files

Commit:
Fixed bug in the code

Commit:
add profile link function

Commit:
include logger line

mr-devs commented on data-loader/database_functions/README.md:
```diff
@@ -0,0 +1,12 @@
+# `database-functions/`
```
Please include a line in the `Contents` section for the `profile_links.py` file.

mr-devs reviewed (approved):
Looks great!
